README.md
CosimR
================

> Generate,modify and save the xml file for co-simualtion between
> external program (R,python or C/C++) and EnergyPlus through the BCVTB

CosimR provides the minimum implementation for co-simulation between
external program (R,python or C/C++) and EnergyPlus through the
SystemCommand Acotr of BCVTB. CosimR can help you to generate(copy),
modify and save a xml file, which will be used in the co-simulation.
Meanwhile, CosimR also provides a simple function to run the
co-simulation through command line of BCVTB.

<img src="man/figures/xml.png"/>

## Installation

You can install the development version of `CosimR` from Github by:

``` r
devtools::install_github('jywang2016/CosimR')
```

## Features

Done:

  - Generate,modify and save BCVTB xml file.
  - Run BCVTB co-simuation based on command line.

Plan:

  - Generate variables.cfg automatically.
  - Check your IDF external interface with the corresponding parameters
    in xml file.

As for more diffcult operations, you need adjust your IDF and external
scripts according to your needs.

## Usage

The flowchart:

``` text

         Step1:Copy                    Step2:Modify         Step3:save/write            Step4: run
xml file =========>>> '.../CosimR.xml' =========>>>  xmlroot =========>>> '.../new.xml' =========>>> result
         copy_xml()                    modify_xml()           write_xml()               run_Cosim()
```

### reproduce the schedule example

We need prepare some files for co-simulation. All files have been placed
in the package. Therefore, just copy them to your target path.

> Note: All the file.copy operations are not necessary for co-simulation
> if you already have your own idf, cfg, R or python script,and so on.
> These copy is just for the reproduction of schedule example which is
> provided in BCVTB.

``` r
library(CosimR)
```

``` r
# This xmlpath is my case. Change it according to your need.
# The idf version is V8-5-0.
xmlpath <- 'D:/bcvtb/examples/CosimRtest'
file.copy(from = system.file("extdata/SmOffPSZ.idf", package = "CosimR"),
          to = xmlpath,
          overwrite = TRUE)
file.copy(from = system.file("extdata/variables.cfg", package = "CosimR"),
          to = xmlpath,
          overwrite = TRUE)
```

### Step1: Copy the xml

``` r
copy_xml(xmlpath = xmlpath)
```

    ## CosimR.xml is copied to D:/bcvtb/examples/CosimRtest successfully!

### Step2: Modify the xml

``` r
xmlroot<- modify_xml(xmlpath = xmlpath,
                     timeStep = 15*60, # '15*60' is more recommended than 15*60
                     beginTime = 3*24*3600,
                     endTime = 4*24*3600,
                     extractLength = 2,
                     idf = paste0(xmlpath,'/SmOffPSZ.idf'),
                     epw = 'D:/bcvtb/examples/ePlusWeather/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw',
                     outname = 'eplusout',
                     programname = 'Rscript',
                     programArgs = 'Cosim.R',
                     workingDir = paste0(xmlpath,'/R'),
                     cpp = FALSE)
```

    ## Part1: Cosim time set modify >>>>>>>>>

    ## timeStep: 900 beginTime: 259200 endTime: 345600

    ## Part2: ArrayExtract actor pram.args modify >>>>>>>>>

    ## extractLength: 2 outputArrayLength: 2

    ## Part3: EnergyPlus Simulator pram.args modify >>>>>>>>>

    ## energyplus command: " -w D:/bcvtb/examples/ePlusWeather/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw -p eplusout -s C -x -m -r D:/bcvtb/examples/CosimRtest/SmOffPSZ.idf "

    ## Part4: SystemCommand pram.args modify >>>>>>>>>

    ## programname: Rscript 
    ## programArgs: "D:/bcvtb/examples/CosimRtest/R/Cosim.R $input $time $iteration" 
    ## workingDir: D:/bcvtb/examples/CosimRtest/R 
    ## simulationLog: D:/bcvtb/examples/CosimRtest/R/simulation.log

    ## ---------------- Modification End ----------------

The `Cosim.R` is also included in the `CosimR` package.

``` r
file.copy(from = system.file("extdata/Cosim.R", package = "CosimR"),
          to = paste0(xmlpath,'/R'),
          overwrite = TRUE)
```

### Step3: write(save) your xml.file

``` r
write_xml(rootnode = xmlroot,
          xmlpath = xmlpath,
          output = 'new.xml')
```

    ## new.xml is written successfully!

**Now, you can operate the new.xml by BCVTB.**

### Step4: run co-simulation (Optional)

``` r
run_Cosim(bcvtbpath = 'D:/bcvtb',
          xmlpath = xmlpath,
          output = 'new.xml')
```

    ## Co-simulation ends. You can check the results inD:/bcvtb/examples/CosimRtest

## result

<img src="man/figures/result.gif" title="Results gif" alt="Results gif" style="display: block; margin: auto;" />

## Python \[optional\]

You can use this part to replace above corresponding R code chunks.

``` r
xmlroot<- modify_xml(xmlpath = xmlpath,
                     timeStep = 15*60, 
                     beginTime = 3*24*3600,
                     endTime = 4*24*3600,
                     extractLength = 2,
                     idf = paste0(xmlpath,'/SmOffPSZ.idf'),
                     epw = 'D:/bcvtb/examples/ePlusWeather/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw',
                     outname = 'eplusout',
                     programname = 'python',
                     programArgs = 'Cosim.py',
                     workingDir = paste0(xmlpath,'/python'),
                     cpp = FALSE)
file.copy(from = system.file("extdata/Cosim.py", package = "CosimR"),
          to = paste0(xmlpath,'/python'),
          overwrite = TRUE)
```

## C/C++ \[optional\]

Use this part to replace above corresponding code chunks.

``` r
xmlroot<- modify_xml(xmlpath = xmlpath,
                     timeStep = 15*60,
                     beginTime = 3*24*3600,
                     endTime = 4*24*3600,
                     extractLength = 2,
                     idf = paste0(xmlpath,'/SmOffPSZ.idf'),
                     epw = 'D:/bcvtb/examples/ePlusWeather/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw',
                     outname = 'eplusout',
                     programname = 'D:/bcvtb/examples/CosimRtest/cpp/Cosim',
                     programArgs = '',
                     workingDir = paste0(xmlpath,'/cpp'),
                     cpp = TRUE)

#window user can use exe
#linux and other OS user can compile the Cosim.cpp
file.copy(from = system.file("extdata/Cosim.exe", package = "CosimR"),
          to = paste0(xmlpath,'/cpp'),
          overwrite = TRUE)
```

## Speed

Run co-simulation for one-day in my windows 10 x64 system:

``` txt
exe(compiled by C++, 5.X secs) >  python(10.x secs) > R(28.x secs).
```

> Calling Rscript by SystemCommand spends much more time than python.

## Author

Jiangyu Wang

*School of Energy and Power Engineering, Huazhong University of Science
and Technology*


NEWS.md
## CosimR 0.1.0

### features

- generate .xml file
- modify .xml file
- save .xml file
- run co-simulation

### scripts

Three templates of external program script are provides to reproduce the `Schedule Example`.

- R 
- python
- C++

R/copy_xml.R
#' copy the xml file to the target directory
#' @description Copy the minimum xml file (CosimR.xml) to the target directory, 
#' for instance, 'C:/bcvtb/examples/your dir/Cosim.xml' in the Windows OS.
#' @param xmlpath The dir you want palce your xml file. Meanwhile, it is also the directory of 
#' your co-simulation project.For me, set the path under the \emph{examples} folder of the BCVTB 
#' installation path is a good choice, namely 'D:/bcvtb/examples/ mydir'.
#' @param overwrite default = TRUE. If there already exists CosimR.xml in your xmlpath, it will
#' be overwritten.
#' @return A xml file. Note: xml_copy function returns no value.
#' @export
copy_xml <- function(xmlpath,overwrite = TRUE){
  if(!dir.exists(xmlpath)){
    message('the xmlpath is created! \n')
    dir.create(xmlpath)
  }
  frompath <- system.file("extdata/CosimR.xml", package = "CosimR")
  status <- file.copy(from = frompath,
                      to = xmlpath,
                      overwrite = overwrite)
  
  if(status){
    message('CosimR.xml is copied to ',xmlpath,' successfully!')
  }else{
    message('XML file copy failed!')
  }
}

R/write_xml.R
#' write the modified xml file to the target directory
#' @description Write the modified xml file to the target directory.
#' @param xmlpath See xmlpath parameters in \code{\link{copy_xml}}
#' @param rootnode The rootnode of xml file. More specifically, the return object of 
#' \code{\link{modify_xml}}
#' @param output new \emph{file name} of your output xml file. For example, output = 'new.xml'.
#' Do not specify the path of your output xml file.
#' @return A new xml file which will be placed in the xmlpath.
#' @import XML
#' @export
write_xml <- function(rootnode,xmlpath,output){
  #tmp <- paste0(tempdir(),'/tmp.xml')
  tmp <- './tmp.xml'
  saveXML(rootnode,file = tmp)
  
  # dtd info parse
  xmlpath2 <- paste0(xmlpath,'/CosimR.xml')
  part1 <- readLines(xmlpath2)
  dtd <- paste0(part1[1:3],collapse = '\n')
  
  # entity info parse
  outputpath <- paste0(xmlpath,'/',output)
  part2 <- suppressWarnings(readLines(tmp))
  file.remove(tmp)
  entity <- paste0(part2,collapse = '\n')
  final <- c(dtd,entity)
  write.table(final,file = outputpath,row.names = FALSE,col.names = FALSE,quote = FALSE)
  
  if(file.exists(outputpath)){
    message(output,' is written successfully!')
  }
}

inst/extdata/Cosim.R
#	This R script is the example of CosimR to show how to do co-simulation with EnergyPlus
#	Package: CosimR
#	Github:	https://github.com/jywang2016/CosimR
#	Author: Jiangyu Wang
#	Date: 11th Dec, 2018
#	Email: jywang_2016@hust.edu.cn
#	Reference: schedule example in BCVTB

args=commandArgs(T)
e  = args[1]
time = args[2]
iter = args[3]

m <- unlist(stringr::str_extract_all(e,'\\d+\\.?\\d*'))
tout = as.numeric(m)[1]

# input const -----------------------
dayStart = 6*3600
nightStart = 16*3600
TCNight = 30
TCRooLow = 22
TCRooHig = 26

TCOutLow = 20
TCOutHig = 24

THDay = 20
THNight = 16

# temp set logic ----------------------
time = as.numeric(time)

ToutDif = tout - TCOutLow

TSetCooOn = TCRooLow + ToutDif * (TCRooHig - TCRooLow)/(TCOutHig - TCOutLow)
TCDay = max(TCRooLow,min(TCRooHig,TSetCooOn))

if(time %% 86400 > dayStart && time %% 86400 < nightStart){
	THSetPoi = THDay
	TCSetPoi = TCDay
}else{
	THSetPoi = THNight
	TCSetPoi = TCNight
}

cat(THSetPoi,',',TCSetPoi,',',sep = '')


inst/extdata/Cosim.py
"""
 	This python script is the example of CosimR to show how to do co-simulation with EnergyPlus
 	Package: CosimR
	Github:	https://github.com/jywang2016/CosimR
	Author: Jiangyu Wang
	Date: 11th Dec, 2018
	Email: jywang_2016@hust.edu.cn
	Reference: schedule example in BCVTB
"""

import re
import sys

string = sys.argv[1]
time = sys.argv[2]
iter = sys.argv[3]

m = re.findall(r"\d+\.?\d*",string)

tout = float(m[0])

# input const -----------------------
dayStart = 6*3600
nightStart = 16*3600
TCNight = 30
TCRooLow = 22
TCRooHig = 26

TCOutLow = 20
TCOutHig = 24

THDay = 20
THNight = 16

# temp set logic ----------------------
time = int(float(time))

ToutDif = tout - TCOutLow

TSetCooOn = TCRooLow + ToutDif * (TCRooHig - TCRooLow)/(TCOutHig - TCOutLow)
TCDay = max(TCRooLow,min(TCRooHig,TSetCooOn))

if time % 86400 > dayStart and time % 86400 < nightStart:
	THSetPoi = THDay
	TCSetPoi = TCDay
else:
	THSetPoi = THNight
	TCSetPoi = TCNight

#print(THSetPoi,',',TCSetPoi,',',sep='')

strs = [str(THSetPoi),str(TCSetPoi),' ']

sys.stdout.write(','.join(strs))


inst/extdata/Cosim.cpp
/*
 This C++ script is the example of CosimR to show how to do co-simulation with EnergyPlus
 The Code is messy because I'm not familiar with the C/C++. 
 You can open a issue and point how to modify it or make it clean.
 Package: CosimR
 Github:	https://github.com/jywang2016/CosimR
 Author: Jiangyu Wang
 Date: 11th Dec, 2018
 Email: jywang_2016@hust.edu.cn
 Reference: schedule example in BCVTB
 Compile: g++ -o Cosim Cosim.cpp -std=c++11
 */
#include <iostream>
#include <string>
#include <regex>
#include <stdio.h>
#include <stdlib.h>

using namespace std;

int main(int argc, char *argv[]){
  
  string error = "the Number of arguments are not correct!";
  if(argc !=4){
    cout<<error<<endl;
    return(1);
  }
  
  smatch result;
  string tempstr = argv[1];
  
  string regex_str("\\d+\\.?\\d*"); 
  regex pat(regex_str,regex::icase);
  
  string::const_iterator iter = tempstr.begin();
  string::const_iterator iterEnd= tempstr.end();
  string strouttemp;
  
  if(regex_search(iter,iterEnd,result,pat)){
    strouttemp = result[0]; //just need the first temperature info
    //cout<<outdoortemp<<endl;
  }
  
  int time;
  char *strtime = (char *) argv[2];
  time = atoi(strtime);
  
  float outtemp;
  char *strtemp = (char *) strouttemp.c_str();
  outtemp = atof(strtemp);
  
  // input const 
  const int DAYSTART = 6*3600;
  const int NIGHTSTART = 16*3600;
  const double TCNIGHT = 30.0;
  const double TCROOLOW = 22.0;
  const double TCROOHIG = 26.0;
  
  const double TCOUTLOW = 20.0;
  const double TCOUTHIG = 24.0;
  
  const double THDAY = 20.0;
  const double THNIGHT = 16.0;
  
  // temp set logic 
  
  double ToutDif;
  ToutDif = outtemp - TCOUTLOW;
  
  double TSetCooOn;
  TSetCooOn = TCROOLOW + ToutDif * (TCROOHIG - TCROOLOW)/(TCOUTHIG - TCOUTLOW);
  
  double TCDay;
  TCDay = max(TCROOLOW,min(TCROOHIG,TSetCooOn));
  
  double THSetPoi;
  double TCSetPoi;
  
  int timeOfDay = time % 86400;
  
  if(timeOfDay > DAYSTART && timeOfDay < NIGHTSTART){
    THSetPoi = THDAY;
    TCSetPoi = TCDay;
  }else{
    THSetPoi = THNIGHT;
    TCSetPoi = TCNIGHT;
  }
  
  cout<<THSetPoi<<","<<TCSetPoi<<", "<<endl;
  
  return 0;
}

R/modify_xml.R
#' Modify the xml according to your Co-simulation requirements.
#' @description Four parts, including Simulation time, ArraryExtract, EnergyPlus simulator and
#' SystemCommand actor, can be modified.
#' @param xmlpath \dots see \code{\link{copy_xml}}
#' @param timeStep default = 15*60 secs. It depends on your timestep parameters in Energyplus 
#' file. By the way, both of the numerical value 15*60 and string "15*60" are supported. But if
#' you want view your xml file, the string form is adviced since it makes you know your step is 
#' 15mins while the numerical value will show confused 900 secs in the BCVTB view.
#' @param beginTime default = 0. Zero means it begin with the date you set in the EnergyPlus 
#' file. Similarly, input a string `'1*24*3600'` is more recommended than a  numerical
#' value 1*24*3600. 
#' @param endTime default = 4*24*3600. It indicates the co-simulation period is 4 days.
#' @param extractLength default = 2. It equals to the variables you want input to the 
#' EnergyPlus.
#' @param idf The idf path. Generally speaking, the idf should be in the root directory of your
#' co-simulation project. Therefore, you can use `paste0(xmlpath,'/SmOffPSZ.idf')` to indicate 
#' your idf path if your idf name is SmOffPSZ.
#' @param epw The weather file path.
#' @param outname default = 'eplusout'. You can use outname to specify the name of messy 
#' simulation results files.
#' @param programname default = 'Rscript'. It means you use \emph{R} exchange data with EnergyPlus.
#' If you set up R path in your environment variable, you can leave it alone. 
#' Or you need specify the complete path of Rscript as 'Your path/R/bin/Rscript'. Similarly,
#' you could set programname as 'python' if you use python to do the co-simulation with 
#' Energyplus. \emph{It is worth noting that you should list the complete path if you use your own 
#' C/C++ based application since there is no environment variable to specify its path.} 
#' @param programArgs 
#' \enumerate{
#'   \item If your programname is 'Rscript', the programArgs should be 'YourScriptName.R'
#'   \item If your programname is 'python', the programArgs should be 'YourScriptName.py'
#'   \item If your programname is 'path/to/your/application', the programArgs should be 
#' "", an empty char.
#' }
#' @param workingDir The subfolder path (usually under your xmlpath) you place your script or 
#' application. If you use R, `paste0(xmlpath,'/R')` is a good choice. For python user, it is
#' `paste0(xmlpath,'/python')`. For cpp user, it is `paste0(xmlpath,'/cpp')`. If there is no 
#' folder path as you specified, the function will create a new one.
#' @param cpp default = FALSE. If the external program is .exe or .sh, which is developed by 
#' yourself, the cpp should be set as \emph{TRUE}
#' @return The rootnode of the modified xml. It will be imported as the rootnode parameter.
#' See \code{\link{write_xml}}.
#' @import XML
#' @export
modify_xml <- function(xmlpath,
                       #timeset
                       timeStep = 15*60,
                       beginTime = 0,
                       endTime = 4*24*3600,
                       #arrayExtract
                       extractLength = 2,
                       #energyplus
                       idf,
                       epw,
                       outname,
                       #syscommand
                       programname = 'Rscript',
                       programArgs,
                       workingDir,
                       cpp = FALSE){
  # read the xml file
  #library(XML)
  xmlpath2 <- paste0(xmlpath,'/CosimR.xml')
  cosim <- xmlParse(xmlpath2)
  rootnode <- xmlRoot(cosim)
  
  # co-simulation time modify
  xmlAttrs(rootnode[[3]])[['value']] <- as.character(timeStep) #timeStep 
  xmlAttrs(rootnode[[4]])[['value']] <- as.character(beginTime) #beginTime
  xmlAttrs(rootnode[[5]])[['value']] <- as.character(endTime) #endTime
  
  message("Part1: Cosim time set modify >>>>>>>>>")
  cat('timeStep:',xmlAttrs(rootnode[[3]])[['value']],
      'beginTime:',xmlAttrs(rootnode[[4]])[['value']],
      'endTime:',xmlAttrs(rootnode[[5]])[['value']],'\n')
  
  # arraryExtract modify
  xmlAttrs(rootnode[[19]][[1]])[['value']] <- as.character(extractLength)
  xmlAttrs(rootnode[[19]][[3]])[['value']] <- as.character(extractLength)
  
  message("Part2: ArrayExtract actor pram.args modify >>>>>>>>>")
  cat('extractLength:',xmlAttrs(rootnode[[19]][[1]])[['value']],
      'outputArrayLength:',xmlAttrs(rootnode[[19]][[3]])[['value']],'\n')
  
  # energyplus simulator modify
  
  message("Part3: EnergyPlus Simulator pram.args modify >>>>>>>>>")
  eplus1 <- paste('-w',epw,'-p',outname,'-s C -x -m -r',idf)
  eplus <- paste('\"',eplus1,'\"')
  xmlAttrs(rootnode[[11]][[2]])[['value']] <- eplus
  cat('energyplus command:',eplus,'\n')
  
  # system command modify
  # if not exists working Dir, then create
  if(!dir.exists(workingDir)){
    cat('the workding dir is created! \n')
    dir.create(workingDir)
  }
  programArgs2 <- paste0('\"',workingDir,'/',programArgs,' $input $time $iteration\"')
  simlog <- paste0(workingDir,'/simulation.log')
  
  # if use .exe in windows, the programArgs should just be programArgs
  # if use .exe in windows, the programname should be added to the end of workdingDir
  if(cpp){
    programname <- paste0(programname)
    programArgs2 <- paste0('\"',programArgs,' $input $time $iteration\"')
  }
  
  xmlAttrs(rootnode[[17]][[1]])[['value']] <- programname #'Rscript'
  xmlAttrs(rootnode[[17]][[2]])[['value']] <- programArgs2 # 
  xmlAttrs(rootnode[[17]][[3]])[['value']] <- workingDir
  xmlAttrs(rootnode[[17]][[4]])[['value']] <- simlog#simulation
  
  message("Part4: SystemCommand pram.args modify >>>>>>>>>")
  cat('programname:',programname,'\n')
  cat('programArgs:',programArgs2,'\n')
  cat('workingDir:',workingDir,'\n')
  cat('simulationLog:',simlog,'\n')
  
  message("---------------- Modification End ----------------")
  # return your modified xml file
  return(rootnode)
}